url79d4f7c7ed_schanq_Live_Algorithm_tgz-pJ8-algorithms_and_systems_AI_Evolutionary_GeneticAlgorithmJ8/src/algorithms_and_systems_AI_Evolutionary/GeneticAlgorithm.java

```diff
@@ -0,0 +1,504 @@
+			if(c.getKeyGene() == k){
+				c.setFitness(4);
+				c.setLevelOfTonality(1);
+				for(Scale s :k.getChord().getScales()){
+					if(c.getScaleGene() == s){
+						c.setFitness(8);
+						c.setLevelOfTonality(2);
+					}
+				}
+				break;
+			}
+		}
+		
+		// Determines whether the Dynamic genes are fit in
+		// comparison to the target chromosome
+		Dynamic [] dyns = this.targetLoudness.allDynamics; 
+		int [] fits     = this.targetLoudness.howCloseAreTheDynamics;
+		DynamicShape ds = this.targetLoudness.getTheShape();
+		
+		for(int i = 0; i < dyns.length; i++){
+			if(dyns[i] ==  c.getDynamicGene()){
+				c.setFitness(c.getFitness()+fits[i]);
+				break;
+			}
+		}
+		if(c.getDshapeGene() == ds){
+			c.setFitness(c.getFitness()+8);
+		}
+		
+		// Determines the Fitness of the Chromosome's rhythmGenes by getting the
+		// beat value of each rhythmic block (considered fit by the target feel) 
+		// and assigning a fitness level for each based on its length so if length 1 then
+		// fitness 2 if beatlength 2 fitness 4 (The purpose is so that the ideal rhythmic 
+		// gene will equate to a fitness of 8). It then compares the chromosome's 
+		// rhythmic_blocks to the ones considered fit, if one matches then add the 
+		// appropriate fitness value to the chromosome's fitness.
+		ArrayList <Rhythmic_Block> fitRhythms = this.targetFeel.getFitRhythms();
+		int [] levelsOfFitness = new int [fitRhythms.toArray().length];
+		for(int i = 0; i < levelsOfFitness.length; i++){
+			if(fitRhythms.get(i).getBeatValue() == 1){
+				levelsOfFitness[i] = 2;
+			}
+			else if(fitRhythms.get(i).getBeatValue() == 2){
+				levelsOfFitness[i] = 4;
+			}
+		}
+		int i = 0;
+		for(Rhythmic_Block rb : fitRhythms){
+			for(Rhythmic_Block r : c.rhythmGenes){
+				if(r.equals(rb)){
+					c.setFitness(c.getFitness()+levelsOfFitness[i]);
+				}
+			}
+			i++;
+		}
+	}
+	// ==========================================================
+	// Genetic crossover - crosses over the genes, however the  =	
+	// specific difference between this and other crossover		=
+	// functions is that it is slightly hard coded so that the 	=
+	// tonal genes (Key and Scale) are crossed over together if =
+	// they are fit.											=
+	// ==========================================================
+	public Chromosome crossover(Chromosome mother, Chromosome father){
+		Chromosome child = new Chromosome();
+		
+		int rand = random.nextInt(2)+1;
+		int randTwoDom = random.nextInt(3);
+		
+		// If neither Chromosome is tonally relevant randomly select either parents' key/scale
+		// to give to the child as all tonal genes are recessive
+		if(mother.getLevelOfTonality() == 0 && father.getLevelOfTonality() == 0){
+			switch(rand){
+			
+			case 1:
+				child = new Chromosome(this.minOctave, this.maxOctave, mother.getKeyGene(), father.getScaleGene());
+				break;
+			case 2:
+				child = new Chromosome(this.minOctave, this.maxOctave, father.getKeyGene(), mother.getScaleGene());
+				break;
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 174:182, 201:209, 242:253.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.